### DISCUSSION TITLE
Install & Build ospos 3,4 Manual – PLEASE help needed

### DISCUSSION THREAD

**Discussion by Giuliano69** [opensourcepos/opensourcepos#4088] [Show and tell]
Hi
I’m trying to jot down a complete instruction on howto install and Build OsPOS 3,4.
The actual build.md seems to me not updated; and some install part are considered obvious (but not for everyone!!) 
I made a fist install&buiold.MD file here.
https://github.com/Giuliano69/opensourcepos/blob/master/ospos%203.4%20install%20%26%20build.md

But it does NOT produce a complete working installation yet :-(( 
When finisched, Last step would be to create an ansible script to make OsPOS install  as simple as a Docker container.

**OPEN PROBLEMS:**
-trying to get The /public page does NOT WORK (CI4 shows a “404 Sorry! Cannot seem to find the page you were looking for. “ Page)
- Instead, logging to localhost give the Login page (!!) 
But WARNS about: _“Welcome to OSPOS!  A database migration to 20240630000001 will start after login. “_
- Logging with normal  user/password (admin-pointofsale), gives _“Whoops!
We seem to have hit a snag. Please try again later…”_

- I noticed that  In the master branch there is no .env file NOR a preload.php file NOR a spark file.
So I cannot run a spark server and try to debut it to find where it “hit the snag”

**Kindly ASK**
- Are the install & Build instruction correct and complete ?
- What may the error be due ?
- What may I test & try ?

Any help would be appreciated !!!

Giuliano

PPS
Is there a chat where we can meet to talk about it ?

BR

**Comment by jekkos** [comment]
Hi you can join the gitter discussion if you like, it might be quicker to discuss issues directly there. I thought the build instructions were quite complete. The .env file is not there indeed it needs to be added manually I believe. The snag you hit is probably the encryption key that is not filed in .env. Perhaps we should describe this as well but typically I do get a clear error message.

**Comment by Giuliano69** [comment]
Hi,
thanks for your reply.

Encription key. 
The install.md of the project talk about it in app/Config/config.php, BUT the new master release simply has NO app/Config/ directory.... nor the Config.php file
HOW to forge the new file if it does not exist (templete?) and WHERE to put it ?

env file. 
Usually I  create the file  from the command line together with the new project: composer create-project codeigniter4/appstarter newporject.... and together with the env file, other php files are  created to let it run by "php spark"    [Pleas Consider  I 've just started using php&CI4  ; I  usually work with other languages...) 
Should I create a FAKE project , and then use tat env file , and those other php loading files ?

public directory:
it states 404 (not found), but the dir is present. Is the virtualhost file I created correct ?
Should the home directory for the website, be var/www/html OR var/www/html/public ?

**Reply by jekkos** [reply]
Ok this might be missing in install instructions. config.php was indeed removed.

**Reply by Giuliano69** [reply]
jekkos, I'm struggling with the 3.4 install..
Tried gitter to contact the team but with no luck... (opensourcepos/Lobby  #opensourcepos_Lobby:gitter.im )

COULD be possible to meet with someone of the team, to manage a correct installation, so that I can produce a correct ansible script ?

Maybe with gitter, telegram o something else...

BR

**Reply by jekkos** [reply]
Yes we can setup a call if you like.I believe bitter has a call functionality no?

**Comment by Giuliano69** [comment]
I tried copying a local .env file into the ospos dir, together with spark and preload.php.
BUT, launching 
``` sh
php spark serve
```
 I get the error below; so it seems that just copying does NOT work .... :
 ```sh
 giuliano@ospos:/var/www/html$ php spark serve
{
    "title": "CodeIgniter\\Cache\\Exceptions\\CacheException",
    "type": "CodeIgniter\\Cache\\Exceptions\\CacheException",
    "code": 500,
    "message": "Cache unable to write to \"/var/www/html/writable/cache/\".",
    "file": "/var/www/html/vendor/codeigniter4/framework/system/Cache/Handlers/FileHandler.php",
    "line": 68,
    "trace": [
        {
            "file": "/var/www/html/vendor/codeigniter4/framework/system/Cache/Handlers/FileHandler.php",
            "line": 68,
            "function": "forUnableToWrite",
            "class": "CodeIgniter\\Cache\\Exceptions\\CacheException",
            "type": "::"
        },
```

Following CI Encryption Service 
 https://codeigniter.com/user_guide/libraries/encryption.html
seems that in CI 4.x the encryption key should not be store in app/Config/config.php, BUT in app/Config/Encription.php.
Is it this the configuration for OSPOS 3.4  ?

**Comment by Giuliano69** [comment]
## .enf file
Digging in writable/logs,
the "hit a snag error" seems related to the lack of .env file.
To me, it seem strange because the master has no .env file, and the CI4 doc seems to suggest the use of NEW files  Config/Database.php and Config/Enfription.ph.
Why this file is still needed ??

```sh
CRITICAL - 2024-10-06 18:46:39 --> ErrorException: copy(/var/www/html/.env): Failed to open stream: No such file or directory
[Method: POST, Route: login]
in APPPATH/Helpers/security_helper.php on line 33.
 1 [internal function]: CodeIgniter\Debug\Exceptions->errorHandler()
 2 APPPATH/Helpers/security_helper.php(33): copy()
 3 APPPATH/Database/Migrations/20220127000000_convert_to_ci4.php(80): check_encryption()
 4 APPPATH/Database/Migrations/20220127000000_convert_to_ci4.php(37): App\Database\Migrations\Convert_to_ci4->convert_ci3_encrypted_data()
 5 SYSTEMPATH/Database/MigrationRunner.php(867): App\Database\Migrations\Convert_to_ci4->up()
 6 SYSTEMPATH/Database/MigrationRunner.php(183): CodeIgniter\Database\MigrationRunner->migrate()
 7 APPPATH/Controllers/Login.php(71): CodeIgniter\Database\MigrationRunner->latest()
 8 SYSTEMPATH/CodeIgniter.php(933): App\Controllers\Login->index()
 9 SYSTEMPATH/CodeIgniter.php(509): CodeIgniter\CodeIgniter->runController()
10 SYSTEMPATH/CodeIgniter.php(355): CodeIgniter\CodeIgniter->handleRequest()
11 SYSTEMPATH/Boot.php(312): CodeIgniter\CodeIgniter->run()
12 SYSTEMPATH/Boot.php(67): CodeIgniter\Boot::runCodeIgniter()
13 FCPATH/index.php(56): CodeIgniter\Boot::bootWeb()
```

Even more strange,  I:
-  copied  a .env file in the app dir, and
- EDITed the file inserting ALL the  DBASE LOGIN in  STANDARD mode,  (tested by a cli shell)
- and set  CI_ENVIRONMENT = development



Well .... I get this database connection error
```sh
Displayed at 22:50:29pm — PHP: 8.3.6 — CodeIgniter: 4.5.1 -- Environment: development
CodeIgniter\Database\Exceptions\DatabaseException

Unable to connect to the database.
Main connection [MySQLi]: Access denied for user '****'@'localhost' (using password: YES) [search →](https://www.duckduckgo.com/?q=CodeIgniter%5CDatabase%5CExceptions%5CDatabaseException+Unable+to+connect+to+the+database.%0AMain+connection+%5BMySQLi%5D%3A+Access+denied+for+user+%40+%28using+password%3A+YES%29)

SYSTEMPATH/Database/BaseConnection.php at line 457
```

Any Idea on where to dig for the solution of the problem  ?

BR
Giuliano
